text/0000-inline-asm.md

```diff
@@ -0,0 +1,892 @@
+
+Both MSVC and D provide what is best described as an embedded DSL for inline assembly.
+It is generally close to the system assembler's syntax, but augmented with the ability to directly access variables that are in scope.
+
+```D
+// This is D code
+int ebx, ecx;
+asm {
+    mov EAX, 4;
+    xor ECX, ECX;
+    cpuid;
+    mov ebx, EBX;
+    mov ecx, ECX;
+}
+writefln("L1 Cache: %s",
+    ((ebx >> 22) + 1) * (((ebx >> 12) & 0x3ff) + 1)
+    * ((ebx & 0xfff) + 1) * (ecx + 1));
+```
+
+```C++
+// This is MSVC C++
+int ebx_v, ecx_v;
+__asm {
+    mov eax, 4
+    xor ecx, ecx
+    cpuid
+    mov ebx_v, ebx
+    mov ecx_v, ecx
+}
+std::cout << "L1 Cache: "
+    << ((ebx_v >> 22) + 1) * (((ebx_v >> 12) & 0x3ff) + 1)
+        * ((ebx_v & 0xfff) + 1) * (ecx_v + 1))
+    << '\n';
+```
+
+While this is very convenient on the user side in that it requires no specification of inputs,
+outputs, or clobbers, it puts a major burden on the implementation.
+The DSL needs to be implemented for each supported architecture, and full knowledge of the
+side-effect of every instruction is required.
+
+This huge implementation overhead is likely one of the reasons MSVC only
+provides this capability for x86, while D at least provides it for x86 and x86-64.
+It should also be noted that the D reference implementation falls slightly short of supporting
+arbitrary assembly. E.g. the lack of access to the `RIP` register makes certain techniques for
+writing position independent code impossible.
+
+As a stop-gap the LDC implementation of D provides a `llvmasm` feature that binds it closely
+to LLVM IR's inline assembly.
+
+We believe it would be unfortunate to put Rust into a similar situation, making certain
+architectures a second-class citizen with respect to inline assembly.
+
+## Provide intrinsics for each instruction
+
+In discussions it is often postulated that providing intrinsics is a better solution to the
+problems at hand.
+However, particularly where precise timing, and full control over the number of generated
+instructions is required intrinsics fall short.
+
+Intrinsics are of course still useful and have their place for inserting specific instructions.
+E.g. making sure a loop uses vector instructions, rather than relying on auto-vectorization.
+
+However, inline assembly is specifically designed for cases where more control is required.
+Also providing an intrinsic for every (potentially obscure) instruction that is needed
+e.g. during early system boot in kernel code is unlikely to scale.
+
+## Make the `asm!` macro return outputs
+
+It has been suggested that the `asm!` macro could return its outputs like the LLVM statement does.
+The benefit is that it is clearer to see that variables are being modified.
+Particular in the case of initialization it becomes more obvious what is happening.
+On the other hand by necessity this splits the direction and constraint specification from
+the variable name, which makes this syntax overall harder to read.
```

Comment:
Instead of returning a tuple, this could return a (locally synthesized, anonymous) struct (I've recently seen a procedural macro do this). That would avoid having to split the variable name from the rest:
```rust
let res = asm!("mul {}", in(reg) a, in("eax") b, lateout("eax") lo, lateout("edx") ho);
res.hi as u64 << 32 + res.lo as u64
```